File: Depth-First Search/Backtracking/Medium/LET_no95_Unique Binary Search Trees II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.helper(1, n)
    
    def helper(self, first, last):
        res = []
        for i in range(first, last + 1):
            for left in self.helper(first, i - 1):
                for right in self.helper(i + 1, last):
                    node = TreeNode(i)
                    node.left = left
                    node.right = right
                    res.append(node)
        return res or [None]

File: Depth-First Search/Backtracking/Medium/LET_no95_Unique Binary Search Trees II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return helper(1, n);
    }
    
    vector<TreeNode*> helper(int first, int last) {
        vector<TreeNode*> res;
        for (int i = first; i <= last; i++) {
            for (auto& left : helper(first, i - 1)) {
                for (auto& right : helper(i + 1, last)) {
                    TreeNode* node = new TreeNode(i);
                    node->left = left;
                    node->right = right;
                    res.push_back(node);
                }
            }
        }
        
        if (!res.empty())
            return res;
        else
            return {nullptr};
    }
};